libraries/AP_Proximity/AP_Proximity_Boundary_3D.h

```diff
@@ -0,0 +1,141 @@
+
+    // stores the layer and sector as a single object to access and modify the 3-D boundary
+    class Face
+    {
+    public:
+
+	    // constructor, invalidate id and distance
+	    Face() { layer = sector = UINT8_MAX; }
+	    Face(uint8_t _layer, uint8_t _sector) { layer = _layer; sector = _sector; }
+
+	    // return true if face has valid layer and sector values
+	    bool valid() const { return ((layer < PROXIMITY_NUM_LAYERS) && (sector < PROXIMITY_NUM_SECTORS)); }
+
+	    // comparison operator
+	    bool operator ==(const Face &other) const { return ((layer == other.layer) && (sector == other.sector)); }
+	    bool operator !=(const Face &other) const { return ((layer != other.layer) || (sector != other.sector)); }
+
+        uint8_t layer;  // vertical "steps" on the 3D Boundary
+        uint8_t sector; // horizontal "steps" on the 3D Boundary
+    };
+
+    // returns face corresponding to the provided yaw and (optionally) pitch
+    // pitch is the vertical body-frame angle (in degrees) to the obstacle (0=directly ahead, 90 is above the vehicle?)
+    // yaw is the horizontal body-frame angle (in degrees) to the obstacle (0=directly ahead of the vehicle, 90 is to the right of the vehicle)
+    Face get_face(float pitch, float yaw) const;
+    Face get_face(float yaw) const { return get_face(0, yaw); }
+
+    // Set the actual body-frame angle(yaw), pitch, and distance of the detected object.
+    // This method will also mark the sector and layer to be "valid",
+    // This distance can then be used for Obstacle Avoidance
+    // Assume detected obstacle is horizontal (zero pitch), if no pitch is passed
+    void set_face_attributes(Face face, float pitch, float yaw, float distance);
+    void set_face_attributes(Face face, float yaw, float distance) { set_face_attributes(face, 0, yaw, distance); }
+
+    // add a distance to the boundary if it is shorter than any other provided distance since the last time the boundary was reset
+    // pitch and yaw are in degrees, distance is in meters
+    void add_distance(float pitch, float yaw, float distance);
+    void add_distance(float yaw, float distance) { add_distance(0, yaw, distance); }
+
+    // update boundary points used for simple avoidance based on a single sector and pitch distance changing
+    //   the boundary points lie on the line between sectors meaning two boundary points may be updated based on a single sector's distance changing
+    //   the boundary point is set to the shortest distance found in the two adjacent sectors, this is a conservative boundary around the vehicle
+    void update_boundary(Face face);
+
+    // update middle layer boundary points
+    void update_middle_boundary();
+
+    // reset boundary.  marks all distances as invalid
+    void reset();
+
+    // Reset this location, specified by Face object, back to default
+    // i.e Distance is marked as not-valid, and set to a large number.
+    void reset_face(Face face);
```

Comment:
Not much for it now on this PR, but it would have been nice to have had `face.reset()`, `face.set_attributes(...)` etc etc.